### ISSUE TITLE
Heavy reporting load can lead in missing samples in StatsdMeterRegistry

### ISSUE THREAD

**Issue by hdv** [micrometer-metrics/micrometer#2908]
**Describe the bug**
Update existing 'receiveParallelMetricsSuccessfully' test with N = 1000 and it starts to fail.

**Environment**
 - Micrometer version 1.7.6 and 1.8.1
 - Micrometer registry StatsdMeterRegistry, Datadog flavor
 - OS: macOS
 - Java version: 
> ~/projects❯ java --version
> openjdk 11.0.2 2019-01-15
> OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
> OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
> 

**To Reproduce**
How to reproduce the bug:
Update existing 'receiveParallelMetricsSuccessfully' test with N = 1000.

**Expected behavior**
Test should not fail with N of any value.

**Additional context**
https://github.com/micrometer-metrics/micrometer/issues/2880

**Comment by marcingrzejszczak** [comment]
Is this still a problem with the latest versions?

**Comment by eager-signal** [comment]
I can confirm that the test still fails, and we’re currently investigating an underreporting issue that seems to be related. In a low-traffic test environment, statsd counters are reported correctly, while in a high traffic production environment, counters seem to be 20–50% underreported.

**Comment by jon-signal** [comment]
I'll also emphasize that, as @hdv noted in https://github.com/micrometer-metrics/micrometer/issues/2880#issuecomment-988287759, this also happens if we change this line: https://github.com/micrometer-metrics/micrometer/blob/f766d261076cd1e04f7dd87fdc7192560b3432b1/implementations/micrometer-registry-statsd/src/test/java/io/micrometer/statsd/StatsdMeterRegistryPublishTest.java#L294

…to:

```java
IntStream.range(0, n).forEach(ignored -> counter.increment());
```

…or, to eliminate any possible doubt:

```java
for (int i = 0; i < n; i++) {
    counter.increment();
}
```

In other words, the problematic behavior here is not restricted to cases where the counter is being updated in parallel.

**Comment by jon-signal** [comment]
Following up on this: we've found and addressed our problem with under-reporting, and it was not related to this issue.

That said, this test certainly fails for any n > 128. I'm not entirely sure why, but for n > 128, the test client will write 128 datagrams (at least for the UDP client), flush, then close. For n <= 128, the test passes because we write all of the datagrams, flush, and close. It's not immediately clear to me why things stop at 128 otherwise. It's also not immediately clear to me if the problem is isolated to this test or if it indicates a real-world problem when using an unbuffered statsd reporter, but I wanted to share what we've learned so far in case it's helpful.

**Comment by eager-signal** [comment]
> It's not immediately clear to me why things stop at 128

I did some additional debugging, and the UDP client is being closed after [`reactor.core.Exceptions$OverflowException: Can't deliver value due to lack of requests`](https://github.com/reactor/reactor-core/blob/fd79952f435fad5f34df3236ae75039a44c3faf0/reactor-core/src/main/java/reactor/core/publisher/SinkManyBestEffort.java#L376). I suspect the 128 value is from [`MonoSend.MAX_SIZE`](https://github.com/reactor/reactor-netty/blob/5f09da87a5008251400c36ee8a2b3b6184d6dd07/reactor-netty-core/src/main/java/reactor/netty/channel/MonoSend.java#L61).

**Comment by marcingrzejszczak** [comment]
I guess we can ask for some help from the Reactor team with this one cc @chemicL

**Comment by chemicL** [comment]
Hey,

as I understand, the test uses an unbuffered configuration, which leads to creating a reactor-netty `UdpClient` that polls a provided `Publisher`. The `Publisher` that the [registry provides](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-statsd/src/main/java/io/micrometer/statsd/StatsdMeterRegistry.java#L228) is a deprecated [DirectProcessor](https://github.com/micrometer-metrics/micrometer/blob/main/implementations/micrometer-registry-statsd/src/main/java/io/micrometer/statsd/StatsdMeterRegistry.java#L91). Each meter at a specified interval pushes observed data to the `Sink` obtained from the `Processor`. In this particular case, an IgnoringSink is used, which ignores overflow and pushes overflowing values down to the `Subscriber`. The `Subscriber` is wrapped in a `DirectInner` class from `SinkManyBestEffort`. The demand comes, as noted above by @eager-signal from the `MonoSendMany`, that provides `SandManyInner` with the `MonoSend.MAX_SIZE` as the refill amount.

The problem lies in the fact that `DirectProcessor` uses `SinkManyBestEffort#directEmitNext` instead of `SinkManyBestEffort#tryEmitNext`.

These are the errors that are ignored and retries are performed to keep reporting:
* UDS: `io.netty.channel.unix.Errors$NativeIoException: writeAddress(..) failed: No buffer space available`
* UDP: `reactor.core.Exceptions$OverflowException: Can't deliver value due to lack of requests`
* TCP: `reactor.core.Exceptions$OverflowException: Can't deliver value due to lack of requests`

Dropping the data seems to be desired, as the strategy is to not buffer, but perhaps that configuration is meant for a different purpose. I'm not too familiar with the use cases here, so I'll leave it to you @marcingrzejszczak to decide what's best. To provide a similar functionality using the new `Sinks` API to replace the `DirectProcessor`, consider the [3.4.0 release notes](https://github.com/reactor/reactor-core/releases/tag/v3.4.0), "Main focus" section.

Other than that, it seems to me that the counter needs to either internally buffer the values so that the polling strategy can be used when there is overflow. Otherwise, an intermediate buffer is needed between the sink and the `UdpClient`/`TcpClient`.
